Update SpiderMonkey

nox opened the pull request:
Do not merge.

highfive commented:
Heads up! This PR modifies the following files:
 * @asajeffrey: components/script/dom/bindings/utils.rs, components/script/dom/eventtarget.rs, components/script/dom/bindings/namespace.rs, components/script/dom/bindings/error.rs, components/script/dom/bindings/codegen/CodegenRust.py and 10 more
 * @kichjang: components/script/dom/bindings/utils.rs, components/script/dom/eventtarget.rs, components/script/dom/bindings/namespace.rs, components/script/dom/bindings/error.rs, components/script/dom/bindings/codegen/CodegenRust.py and 10 more
 * @emilio: components/style/Cargo.toml

highfive commented:
<img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20> **Warning** <img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20>

* These commits modify **unsafe code**. Please review it carefully!
* These commits modify style and script code, but no tests are modified. Please consider adding a test!

nox commented:
@bors-servo try=wpt

bors-servo commented:
:hourglass: Trying commit 43e811c256a1832de0455009f80f7c1f061a507c with merge 3c3b00aefe84b7cfcfccf54f1bf27471a615d67d...

bors-servo commented:
:broken_heart: Test failed - [status-taskcluster](https://community-tc.services.mozilla.com/tasks/groups/bv0qxhrnQF-Yl7cKMgTPrA)

nox commented:
Right now I'm looking into this crash:

```
  ▶ CRASH [expected OK] /webmessaging/event.data.sub.htm
  │ 
  │ 
  │ 
  │ 
  │ Stack trace for thread "ScriptThread PipelineId { namespace_id: PipelineNamespaceId(1), index: PipelineIndex(1) }"
  │ stack backtrace:
  │    0: servo::backtrace::print
  │    1: servo::install_crash_handler::handler
  │    2: _ZL15WasmTrapHandleriP9siginfo_tPv
  │              at /root/.cargo/git/checkouts/mozjs-fa11ffc7d4f1cc2d/554d944/mozjs/js/src/wasm/WasmSignalHandlers.cpp:970
  │    3: <unknown>
  │    4: _ZNK2JS14ContextOptions13extraWarningsEv
  │              at /repo/target/release/build/mozjs_sys-fdad495e57b1b58d/out/build/dist/include/js/ContextOptions.h:149
  │       _ZNK2JS14RealmBehaviors13extraWarningsEP9JSContext
  │              at /root/.cargo/git/checkouts/mozjs-fa11ffc7d4f1cc2d/554d944/mozjs/js/src/jsapi.cpp:1665
  │       CompileOptions
  │              at /root/.cargo/git/checkouts/mozjs-fa11ffc7d4f1cc2d/554d944/mozjs/js/src/jsapi.cpp:3560
  │    5: NewCompileOptions
  │    6: _ZN6script3dom11eventtarget19InlineEventListener20get_compiled_handler17h5b1ca2c7e4d5dca8E.llvm.4617923786188796215
  │    7: <alloc::vec::Vec<T> as alloc::vec::SpecExtend<T,I>>::from_iter
  │    8: script::dom::eventtarget::EventTarget::get_listeners_for
  │    9: script::dom::event::invoke
  │   10: script::dom::event::Event::dispatch
  │   11: script::dom::eventtarget::EventTarget::fire_event_with_params
  │   12: script::dom::htmliframeelement::HTMLIFrameElement::iframe_load_event_steps
  │   13: script::script_thread::ScriptThread::handle_msg_from_constellation
```

There are many others though.

nox commented:
I found that fix. Retrying now.

@bors-servo try=wpt

bors-servo commented:
:hourglass: Trying commit 6e18c217208aef822593546ed828f83e088e5fea with merge 1136e8b7b2bd128385cdfa3b24bb7461f93600ea...

bors-servo commented:
:broken_heart: Test failed - [status-taskcluster](https://community-tc.services.mozilla.com/tasks/groups/FFknarIESFyBIoEjwB2S4w)

nox commented:
```
 0:10.15 pid:68126    0: <servo::backtrace::Print as core::fmt::Debug>::fmt
 0:10.16 pid:68126    1: core::fmt::write
 0:10.16 pid:68126    2: <std::io::stdio::Stdout as std::io::Write>::write_fmt
 0:10.16 pid:68126    3: std::io::stdio::_print
 0:10.16 pid:68126    4: servo::backtrace::print
 0:10.16 pid:68126    5: servo::install_crash_handler::handler
 0:10.16 pid:68126    6: _sigtramp
 0:10.16 pid:68126    7: _ZN2js13ContextChecks4failEPN2JS11CompartmentES3_i
 0:10.16 pid:68126    8: _ZN2js13ContextChecks5checkEPN2JS11CompartmentEi
 0:10.16 pid:68126    9: _ZN2js13ContextChecks5checkEP8JSObjecti
 0:10.16 pid:68126   10: _ZN9JSContext9checkImplIN2JS6HandleIP8JSObjectEEJNS2_INS1_11PropertyKeyEEEEEEviRKT_DpRKT0_
 0:10.16 pid:68126   11: _ZN9JSContext5checkIJN2JS6HandleIP8JSObjectEENS2_INS1_11PropertyKeyEEEEEEvDpRKT_
 0:10.16 pid:68126   12: _Z18JS_HasPropertyByIdP9JSContextN2JS6HandleIP8JSObjectEENS2_INS1_11PropertyKeyEEEPb
 0:10.17 pid:68126   13: _Z14JS_HasPropertyP9JSContextN2JS6HandleIP8JSObjectEEPKcPb
 0:10.17 pid:68126   14: script::dom::bindings::conversions::get_property_jsval
 0:10.17 pid:68126   15: script::webdriver_handlers::handle_get_property
 0:10.17 pid:68126   16: script::script_thread::ScriptThread::handle_webdriver_msg
 0:10.17 pid:68126   17: script::script_thread::ScriptThread::handle_msg_from_constellation
 0:10.17 pid:68126   18: _ZN6script13script_thread12ScriptThread11handle_msgs17h733b1905b3ce12e6E.llvm.342661483298543719
 0:10.17 pid:68126   19: profile_traits::mem::ProfilerChan::run_with_memory_reporting
 0:10.17 pid:68126   20: _ZN3std10sys_common9backtrace28__rust_begin_short_backtrace17h5dbc6f21461c3103E.llvm.17006953112752071020
 0:10.17 pid:68126   21: _ZN3std9panicking3try7do_call17hbe656bbf98c2bfebE.llvm.8896023105422182164
 0:10.17 pid:68126   22: __rust_maybe_catch_panic
 0:10.18 pid:68126   23: core::ops::function::FnOnce::call_once{{vtable.shim}}
 0:10.18 pid:68126   24: <alloc::boxed::Box<F> as core::ops::function::FnOnce<A>>::call_once
 0:10.18 pid:68126   25: std::sys::unix::thread::Thread::new::thread_start
 0:10.18 pid:68126   26: _pthread_body
 0:10.18 pid:68126   27: _pthread_start
```

nox commented:
Fixed this one too. Seems like a *lot* of the errors are missing `JSAutoCompartment` uses, which is absolutely terrifying.

nox commented:
@bors-servo try=wpt

bors-servo commented:
:hourglass: Trying commit eca6c359612e8db682a6163fa5b9eacaec5b97d7 with merge 3bea26461721b2091103466afa525fe6b5241a8c...

bors-servo commented:
:broken_heart: Test failed - [status-taskcluster](https://community-tc.services.mozilla.com/tasks/groups/QuvVue-MR2Guxjc8UaqtZw)

nox commented:
This build error makes no sense and I'll pretend like I didn't see anything.

@bors-servo try=wpt

bors-servo commented:
:hourglass: Trying commit 315e8a1451eb4d4d189e71c30ed365545ee79dd5 with merge ff3979bdf621e83ac295e85ccaf9e35d226942d0...

nox commented:
I'm trying to understand this one:

```
 0:08.33 pid:6997 Stack trace for thread "ScriptThread PipelineId { namespace_id: PipelineNamespaceId(1), index: PipelineIndex(1) }"
 0:08.33 pid:6997 stack backtrace:
 0:08.36 pid:6997    0: <servo::backtrace::Print as core::fmt::Debug>::fmt
 0:08.36 pid:6997    1: core::fmt::write
 0:08.36 pid:6997    2: std::io::Write::write_fmt
 0:08.36 pid:6997    3: servo::backtrace::print
 0:08.36 pid:6997    4: servo::install_crash_handler::handler
 0:08.36 pid:6997    5: _sigtramp
 0:08.36 pid:6997    6: <std::panic::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once
 0:08.36 pid:6997    7: _ZN3std9panicking3try7do_call17hed4982c729e2a316E.llvm.8313466446934261671
 0:08.37 pid:6997    8: __rust_maybe_catch_panic
 0:08.37 pid:6997    9: script::dom::bindings::codegen::Bindings::WindowBinding::WindowBinding::get_self
```

This is not about realms and it is pretty scary.

nox commented:
A debug build would probably help, but #25550.

nox commented:
@bors-servo try=wpt

bors-servo commented:
:hourglass: Trying commit 5238567c4e4f5e2d4042efc9141072f66503ca86 with merge 962d8075cc58fca88fda978383fcd77a1f10974a...

bors-servo commented:
:broken_heart: Test failed - [status-taskcluster](https://community-tc.services.mozilla.com/tasks/groups/dNB54DmuRkex3KlZepu-gA)

nox commented:
Enabling gczeal, I get this:

```
 0:02.64 pid:15461 assertion failed: `(left != right)`
 0:02.64 pid:15461   left: `0`,
 0:02.64 pid:15461  right: `0` (thread ScriptThread PipelineId { namespace_id: PipelineNamespaceId(1), index: PipelineIndex(1) }, at components/script/dom/windowproxy.rs:156)
 0:02.64 pid:15461 stack backtrace:
 0:02.67 pid:15461    0: <servo::backtrace::Print as core::fmt::Debug>::fmt
 0:02.67 pid:15461    1: core::fmt::write
 0:02.67 pid:15461    2: std::io::Write::write_fmt
 0:02.67 pid:15461    3: servo::backtrace::print
 0:02.67 pid:15461    4: servo::main::{{closure}}
 0:02.67 pid:15461    5: std::panicking::rust_panic_with_hook
 0:02.67 pid:15461    6: rust_begin_unwind
 0:02.67 pid:15461    7: std::panicking::begin_panic_fmt
 0:02.67 pid:15461    8: script::dom::windowproxy::WindowProxy::new
 0:02.67 pid:15461    9: script::script_thread::ScriptThread::local_window_proxy
 0:02.68 pid:15461   10: script::script_thread::ScriptThread::load
 0:02.68 pid:15461   11: script::script_thread::ScriptThread::handle_page_headers_available
 0:02.68 pid:15461   12: std::thread::local::LocalKey<T>::with
 0:02.68 pid:15461   13: <script::dom::servoparser::ParserContext as net_traits::FetchResponseListener>::process_response
 0:02.68 pid:15461   14: script::script_thread::ScriptThread::handle_msg_from_constellation
 0:02.68 pid:15461   15: _ZN6script13script_thread12ScriptThread11handle_msgs17h2e83c538c2024641E.llvm.11052331390567343350
 0:02.68 pid:15461   16: profile_traits::mem::ProfilerChan::run_with_memory_reporting
 0:02.68 pid:15461   17: _ZN3std10sys_common9backtrace28__rust_begin_short_backtrace17hbab79c8768f38f56E.llvm.9041608547744892394
 0:02.68 pid:15461   18: _ZN3std9panicking3try7do_call17h588435982eaedc55E.llvm.10582130529027147413
 0:02.69 pid:15461   19: __rust_maybe_catch_panic
 0:02.69 pid:15461   20: core::ops::function::FnOnce::call_once{{vtable.shim}}
 0:02.69 pid:15461   21: <alloc::boxed::Box<F> as core::ops::function::FnOnce<A>>::call_once
 0:02.69 pid:15461   22: std::sys::unix::thread::Thread::new::thread_start
 0:02.69 pid:15461   23: _pthread_body
 0:02.69 pid:15461   24: _pthread_start
 0:02.69 pid:15461 [2020-02-05T15:21:27Z ERROR servo] assertion failed: `(left != right)`
 0:02.69 pid:15461       left: `0`,
 0:02.69 pid:15461      right: `0`
 0:02.71 pid:15461 Pipeline failed in hard-fail mode.  Crashing!
```

I checked the name of that `JSClass` and it was just "Object", which means something is completely F'd somewhere.

nox commented:
This "Object" mess happens on master too.

Currently trying to see if maybe our `JSJitInfo` bindings are outdated in some way, but I feel like Servo would break way more if that was the case.

nox commented:
The `JSJitInfo` struct looks correct to me.

nox commented:
So, some progress: I can't reproduce failures if I set "js.ion.enabled" to false. Obviously that's not the right fix, but it gives some clue as to what the hell is happening.

nox commented:
@bors-servo try=wpt

bors-servo commented:
:hourglass: Trying commit 1beab1a3b753fa6aaa0fd2500040835e648f3505 with merge 0247efe3f54323f399f50b78d35102231ae05be0...

bors-servo commented:
:broken_heart: Test failed - [status-taskcluster](https://community-tc.services.mozilla.com/tasks/groups/ILkxhIt7QQCvGv91PxO-BA)

nox commented:
It seems like nothing fails anymore if I disable ion, but obviously I can't land that…

nox commented:
With ion disabled:

```
  ▶ OK [expected TIMEOUT] /html/semantics/scripting-1/the-script-element/css-module/css-module-worker-test.html
  ▶ Unexpected subtest result in /html/semantics/scripting-1/the-script-element/css-module/css-module-worker-test.html:
  └ PASS [expected TIMEOUT] A dynamic import CSS Module within a web worker should not load.
  ▶ Unexpected subtest result in /html/semantics/scripting-1/the-script-element/module/inactive-context-import.html:
  └ PASS [expected FAIL] dynamic import from inactive context should not crash
  ▶ OK [expected ERROR] /html/infrastructure/safe-passing-of-structured-data/structured_clone_bigint.html
  ▶ Unexpected subtest result in /_mozilla/mozilla/interfaces.worker.html:
  │ FAIL [expected PASS] interfaces
  │   → assert_true: If this is failing: DANGER, are you sure you want to expose the new interface BigInt64Array to all webpages as a property on the global? Do not make a change to this file without review from jdm or Ms2ger for that specific change! expected true got false
  │ 
  │ test_interfaces/<@http://web-platform.test:8000/_mozilla/mozilla/interfaces.js:72:18
  │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:2024:25
  │ test@http://web-platform.test:8000/resources/testharness.js:548:30
  │ test_interfaces@http://web-platform.test:8000/_mozilla/mozilla/interfaces.js:2:7
  └ @http://web-platform.test:8000/_mozilla/mozilla/interfaces.worker.js:9:16
  ▶ Unexpected subtest result in /_mozilla/mozilla/localeCompare.html:
  │ FAIL [expected PASS] localeCompare should return the same as other browsers, even though it's implementation-dependent
  │   → assert_equals: expected -1 but got 1
  │ 
  │ @http://web-platform.test:8000/_mozilla/mozilla/localeCompare.html:8:16
  │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:2024:25
  │ test@http://web-platform.test:8000/resources/testharness.js:548:30
  └ @http://web-platform.test:8000/_mozilla/mozilla/localeCompare.html:7:5
  ▶ Unexpected subtest result in /_mozilla/mozilla/interfaces.html:
  │ FAIL [expected PASS] Interfaces exposed on the window
  │   → assert_true: If this is failing: DANGER, are you sure you want to expose the new interface BigInt64Array to all webpages as a property on the global? Do not make a change to this file without review from jdm or Ms2ger for that specific change! expected true got false
  │ 
  │ test_interfaces/<@http://web-platform.test:8000/_mozilla/mozilla/interfaces.js:72:18
  │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:2024:25
  │ test@http://web-platform.test:8000/resources/testharness.js:548:30
  │ test_interfaces@http://web-platform.test:8000/_mozilla/mozilla/interfaces.js:2:7
  └ @http://web-platform.test:8000/_mozilla/mozilla/interfaces.html:13:16
```

The only one that seems weird to me is "localeCompare should return the same as other browsers, even though it's implementation-dependent".

nox commented:
@bors-servo try=wpt

bors-servo commented:
:hourglass: Trying commit e26cdbd821f65167304813246ba0a3c3851e2cfe with merge 20129e2c77be770cbc3685576057a3c1381a906b...

bors-servo commented:
:broken_heart: Test failed - [status-taskcluster](https://community-tc.services.mozilla.com/tasks/groups/IV4otZqiTB24IkVD-Oz21g)

nox commented:
@bors-servo try=wpt

bors-servo commented:
:hourglass: Trying commit ba036fecb98e0f1c5f8140adc17454859b305843 with merge 111460558c866c6fcd367f03a71caa8b10de184d...

bors-servo commented:
:sunny: Test successful - [status-taskcluster](https://community-tc.services.mozilla.com/tasks/groups/Ceu2KwjARU2TwB8-xm8weA)
State: approved= try=True

nox commented:
Let it be known that I'm extremely confused that this is now all green. I don't know what I changed.

nox commented:
Seems like tests pass in debug mode, and they don't in release mode. Great.

nox commented:
I did manage to get a failure locally with a better stacktrace, but it turns out this was not helpful either:

```
± ./mach test-wpt --no-pause /_webgl/conformance/attribs/gl-vertexattribpointer-offsets.html
 0:01.38 INFO Using 12 client processes
 0:01.43 INFO Starting http server on 127.0.0.1:8001
 0:01.43 INFO Starting http server on 127.0.0.1:8000
 0:01.44 INFO Starting https server on 127.0.0.1:8443
 0:01.93 SUITE_START: web-platform-test - running 1 tests
 0:01.93 INFO Running reftest tests
 0:01.93 INFO No reftest tests to run
 0:01.93 INFO Running wdspec tests
 0:01.93 INFO No wdspec tests to run
 0:01.93 INFO Running testharness tests
 0:01.94 INFO No more tests
 0:01.94 INFO Starting runner
 0:01.94 INFO No more tests
 0:01.94 INFO No more tests
 0:01.94 INFO No more tests
 0:01.94 INFO No more tests
 0:01.94 INFO No more tests
 0:01.95 INFO No more tests
 0:01.95 INFO No more tests
 0:01.95 INFO No more tests
 0:01.95 INFO No more tests
 0:01.95 INFO No more tests
 0:01.95 TEST_START: /_webgl/conformance/attribs/gl-vertexattribpointer-offsets.html
 0:01.95 WARNING Got command init_succeeded in state running
 0:02.18 pid:69252 Full command: /Users/nox/src/servo/target/debug/servo --hard-fail -u Servo/wptrunner -z http://web-platform.test:8000/_webgl/conformance/attribs/gl-vertexattribpointer-offsets.html -Z replace-surrogates --user-stylesheet /Users/nox/src/servo/resources/ahem.css --pref dom.offscreen_canvas.enabled=true --certificate-path /Users/nox/src/servo/tests/wpt/web-platform-tests/tools/certs/cacert.pem
pid:69252 
 0:02.18 pid:69252 
 0:02.18 pid:69252 
 0:04.78 pid:69252 Stack trace for thread "ScriptThread PipelineId { namespace_id: PipelineNamespaceId(1), index: PipelineIndex(1) }"
 0:04.78 pid:69252 stack backtrace:
 0:04.93 pid:69252    0: backtrace::backtrace::trace_unsynchronized
 0:04.94 pid:69252    1: <servo::backtrace::Print as core::fmt::Debug>::fmt
 0:04.94 pid:69252    2: core::fmt::write
 0:04.95 pid:69252    3: std::io::Write::write_fmt
 0:04.95 pid:69252    4: servo::backtrace::print
 0:04.96 pid:69252    5: servo::install_crash_handler::handler
 0:04.96 pid:69252    6: _sigtramp
 0:04.97 pid:69252    7: core::cell::Cell<T>::get
 0:04.97 pid:69252    8: core::cell::BorrowRef::new
 0:04.98 pid:69252    9: core::cell::RefCell<T>::try_borrow
 0:04.99 pid:69252   10: script::dom::bindings::cell::DomRefCell<T>::borrow
 0:04.99 pid:69252   11: script::dom::webgl_extensions::extensions::WebGLExtensions::is_get_parameter_name_enabled
 0:05.00 pid:69252   12: <script::dom::webglrenderingcontext::WebGLRenderingContext as script::dom::bindings::codegen::Bindings::WebGLRenderingContextBinding::WebGLRenderingContextBinding::WebGLRenderingContextMethods>::GetParameter
 0:05.01 pid:69252   13: script::dom::bindings::codegen::Bindings::WebGLRenderingContextBinding::WebGLRenderingContextBinding::getParameter::{{closure}}
 0:05.01 pid:69252   14: core::ops::function::FnOnce::call_once
 0:05.02 pid:69252   15: <std::panic::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once
 0:05.03 pid:69252   16: std::panicking::try::do_call
 0:05.04 pid:69252   17: __rust_maybe_catch_panic
 0:05.05 pid:69252   18: std::panicking::try
 0:05.05 pid:69252   19: std::panic::catch_unwind
 0:05.06 pid:69252   20: mozjs::panic::wrap_panic
 0:05.07 pid:69252   21: script::dom::bindings::codegen::Bindings::WebGLRenderingContextBinding::WebGLRenderingContextBinding::getParameter
 0:05.13 TEST_END: CRASH, expected OK
 0:05.13 INFO No more tests
 0:05.14 INFO Closing logging queue
 0:05.14 INFO queue closed
 0:05.14 INFO Running crashtest tests
 0:05.14 INFO No crashtest tests to run
 0:05.14 INFO Got 1 unexpected results
 0:05.14 SUITE_END

web-platform-test
~~~~~~~~~~~~~~~~~
Ran 1 checks (1 tests)
Expected results: 0
Unexpected results: 1
  test: 1 (1 crash)

Unexpected Results
------------------
CRASH /_webgl/conformance/attribs/gl-vertexattribpointer-offsets.html
 0:05.16 INFO Closing logging queue
 0:05.16 INFO queue closed
```

nox commented:
Running the test under gdb just makes it hang forever.

nox commented:
Other things I did trying to debug that infuriating smup:

* Triple-checked in [rust-mozjs](https://github.com/servo/rust-mozjs/pull/494/files?w=1) that my changes to jsglue.cpp and glue.rs were kept in sync
* Checked that my AutoIdVector changes (in rust-mozjs again) make sense
* Checked that we correctly use JSJitInfo

jdm commented:
I ran /css/css-transforms/animation/transform-interpolation-001.html under gdb in a release+debugmozjs build on linux. It's crashing in WindowBinding::get_window when invoking `this.Window()`; I tried setting a breakpoint in the _finalize method in that file and it's not getting hit, so we're getting an invalid pointer as `this` when being invoked by the JIT.

nox commented:
My GC zeal fix does not solve the problem.

nox commented:
@bors-servo try=wpt

Redoing a try because I want to try and find a simpler test that fails for another interface than Window.

bors-servo commented:
:hourglass: Trying commit 75d547205b71a7470dff468d8e764976ef1cc751 with merge 67df506dd9b4ea5f48a3d7f08a1bf00ff49f0af2...

bors-servo commented:
:broken_heart: Test failed - [status-taskcluster](https://community-tc.services.mozilla.com/tasks/groups/XNxvgocMSyyqB8ywx1Mlng)

nox commented:
@bors-servo try=wpt

bors-servo commented:
:hourglass: Trying commit a9fd127ed07e76283cb11fc9cd6707b29de7af94 with merge 4d73731aba8f7f2d1890c4882422ce392cdffd97...

nox commented:
I've managed to make two smaller cases that aren't tests:

```html
<img id=img>
<script>
  var img = document.getElementById('img');
  while (true) {
    img.complete
  }
</script>
```

``
   0: backtrace::backtrace::trace_unsynchronized
   1: <servo::backtrace::Print as core::fmt::Debug>::fmt
   2: core::fmt::write
   3: std::io::Write::write_fmt
   4: servo::backtrace::print
   5: servo::install_crash_handler::handler
   6: _sigtramp
   7: core::cell::Cell<T>::get
   8: core::cell::BorrowRef::new
   9: core::cell::RefCell<T>::try_borrow
  10: script::dom::bindings::cell::DomRefCell<T>::borrow
  11: script::dom::element::Element::has_attribute
  12: <script::dom::htmlimageelement::HTMLImageElement as script::dom::bindings::codegen::Bindings::HTMLImageElementBinding::HTMLImageElementBinding::HTMLImageElementMethods>::Complete
  13: script::dom::bindings::codegen::Bindings::HTMLImageElementBinding::HTMLImageElementBinding::get_complete::{{closure}}
  14: core::ops::function::FnOnce::call_once
  15: <std::panic::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once
  16: std::panicking::try::do_call
  17: __rust_maybe_catch_panic
  18: std::panicking::try
  19: std::panic::catch_unwind
  20: mozjs::panic::wrap_panic
  21: script::dom::bindings::codegen::Bindings::HTMLImageElementBinding::HTMLImageElementBinding::get_complete
```

```html
<body>
<script>
  while (true) {
    document.body.text
  }
</script>
```

```
   0: backtrace::backtrace::trace_unsynchronized
   1: <servo::backtrace::Print as core::fmt::Debug>::fmt
   2: core::fmt::write
   3: std::io::Write::write_fmt
   4: servo::backtrace::print
   5: servo::install_crash_handler::handler
   6: _sigtramp
```

bors-servo commented:
:broken_heart: Test failed - [status-taskcluster](https://community-tc.services.mozilla.com/tasks/groups/OkUIqGvkQQyZrNA82vgz0g)

nox commented:
My second sample does that on your box under gdb, @jdm:

```
(gdb) bt
#0  0x00007ffff371a57e in __memmove_sse2_unaligned_erms ()
    at ../sysdeps/x86_64/multiarch/memmove-vec-unaligned-erms.S:299
#1  0x00005555586727f0 in core::intrinsics::copy_nonoverlapping (src=0x3fffed5abf60, dst=0x7fff91157048, count=1)
    at /rustc/3291ae33907f2a866ea6cea89113200555038d06/src/libcore/intrinsics.rs:1523
#2  0x00005555575e4c60 in core::ptr::read (src=0x3fffed5abf60)
    at /rustc/3291ae33907f2a866ea6cea89113200555038d06/src/libcore/ptr/mod.rs:627
#3  0x00005555577fd6ae in script::dom::bindings::root::MutNullableDom<T>::get (self=0x3fffed5abf60)
    at components/script/dom/bindings/root.rs:565
#4  0x0000555558cf23cd in <script::dom::window::Window as script::dom::bindings::codegen::Bindings::WindowBinding::WindowBinding::WindowMethods>::Document (self=0x3fffed5ab800) at components/script/dom/window.rs:769
#5  0x0000555559bac6a8 in script::dom::bindings::codegen::Bindings::WindowBinding::WindowBinding::get_document::{{closure}} () at /home/jdm/servo-smup/target/debug/build/script-1d8a7ae9da282fe5/out/Bindings/WindowBinding.rs:1041
#6  0x0000555559599449 in core::ops::function::FnOnce::call_once ()
    at /rustc/3291ae33907f2a866ea6cea89113200555038d06/src/libcore/ops/function.rs:232
#7  0x0000555559f74682 in <std::panic::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once (self=..., _args=()) at /rustc/3291ae33907f2a866ea6cea89113200555038d06/src/libstd/panic.rs:318
#8  0x000055555ad41d23 in std::panicking::try::do_call (data=0x7fff91157378 "\370t\025\221\377\177")
    at /rustc/3291ae33907f2a866ea6cea89113200555038d06/src/libstd/panicking.rs:305
#9  0x0000555560fe8ab7 in __rust_maybe_catch_panic () at src/libpanic_unwind/lib.rs:86
#10 0x000055555aa3ccd6 in std::panicking::try (f=...)
    at /rustc/3291ae33907f2a866ea6cea89113200555038d06/src/libstd/panicking.rs:281
#11 0x0000555559fcac49 in std::panic::catch_unwind (f=...)
    at /rustc/3291ae33907f2a866ea6cea89113200555038d06/src/libstd/panic.rs:394
#12 0x0000555557b27058 in mozjs::panic::wrap_panic (function=..., generic_return_type=false)
    at /home/jdm/.cargo/git/checkouts/rust-mozjs-8611526964119dd6/95b07b4/src/panic.rs:22
#13 0x000055555a3c8a7b in script::dom::bindings::codegen::Bindings::WindowBinding::WindowBinding::get_document (cx=0x7fff7000ea70, _obj=..., this=0x3fffed5ab800, args=...)
    at /home/jdm/servo-smup/target/debug/build/script-1d8a7ae9da282fe5/out/Bindings/WindowBinding.rs:1037
#14 0x000000b344e450a5 in  ()
#15 0x0000000000000000 in  ()
```

bors-servo commented:
:umbrella: The latest upstream changes (presumably #25780) made this pull request unmergeable. Please resolve the merge conflicts.

jdm commented:
I am 99% sure this problem is happening because our JSValue creation and extraction code no longer matches the C++ methods. Compare https://searchfox.org/mozilla-central/source/js/public/Value.h#798-812 against:
* https://github.com/servo/mozjs/blob/master/src/jsval.rs#L182-L196
* https://github.com/servo/mozjs/blob/bc4935b668171863e537d3fb0d911001a6742013/src/jsval.rs#L436-L449
* https://github.com/servo/rust-mozjs/blob/master/src/jsval.rs#L181-L195

The invalid pointer that I observe when the segfault occurs matches the valid pointer shifted right one bit. The reason this only affects the JIT is that all of our Rust code using `.toPrivate()` performs the shifting that we did when creating values with `PrivateValue(..)`, but the JIT relies on the C++ implementations which do not.

jdm commented:
I think the best way to avoid this problem in the future is to add C wrappers for all of the C++ JSValue inline creation/extraction methods, and write a unit test that verifies that our Rust implementations match for all of the possible cases.

jdm commented:
I verified that removing the shifting from those rust implementations makes the testcase no longer crash. I'm not planning to keep looking into it at this point.

nox commented:
Yeah I'm fairly confident this was all that was needed. Pretty disappointing given that's literally the one piece of code I didn't check got out of sync with the C++ code. I will also file a bug upstream that their own jsval.rs is outdated too.

nox commented:
@bors-servo try=wpt

bors-servo commented:
:hourglass: Trying commit 8b780426759bfd6e6d2d26ec62437365a12f2714 with merge e14e1606bcecf0bd6160481df9c02a9aef6455d2...

bors-servo commented:
:broken_heart: Test failed - [status-taskcluster](https://community-tc.services.mozilla.com/tasks/groups/es1joUiDRpezsflBKZMvXA)

nox commented:
```
  ▶ OK [expected TIMEOUT] /html/semantics/scripting-1/the-script-element/css-module/css-module-worker-test.html
  ▶ Unexpected subtest result in /html/semantics/scripting-1/the-script-element/css-module/css-module-worker-test.html:
  └ PASS [expected TIMEOUT] A dynamic import CSS Module within a web worker should not load.

▶ Unexpected subtest result in /html/semantics/scripting-1/the-script-element/module/inactive-context-import.html:
  └ PASS [expected FAIL] dynamic import from inactive context should not crash

  ▶ OK [expected ERROR] /html/infrastructure/safe-passing-of-structured-data/structured_clone_bigint.html

  ▶ Unexpected subtest result in /_mozilla/mozilla/interfaces.worker.html:
  │ FAIL [expected PASS] interfaces
  │   → assert_true: If this is failing: DANGER, are you sure you want to expose the new interface BigUint64Array to all webpages as a property on the global? Do not make a change to this file without review from jdm or Ms2ger for that specific change! expected true got false
  │ 
  │ test_interfaces/<@http://web-platform.test:8000/_mozilla/mozilla/interfaces.js:73:18
  │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:2024:25
  │ test@http://web-platform.test:8000/resources/testharness.js:548:30
  │ test_interfaces@http://web-platform.test:8000/_mozilla/mozilla/interfaces.js:2:7
  └ @http://web-platform.test:8000/_mozilla/mozilla/interfaces.worker.js:9:16
  ▶ Unexpected subtest result in /_mozilla/mozilla/localeCompare.html:
  │ FAIL [expected PASS] localeCompare should return the same as other browsers, even though it's implementation-dependent
  │   → assert_equals: expected -1 but got 1
  │ 
  │ @http://web-platform.test:8000/_mozilla/mozilla/localeCompare.html:8:16
  │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:2024:25
  │ test@http://web-platform.test:8000/resources/testharness.js:548:30
  └ @http://web-platform.test:8000/_mozilla/mozilla/localeCompare.html:7:5
  ▶ Unexpected subtest result in /_mozilla/mozilla/interfaces.html:
  │ FAIL [expected PASS] Interfaces exposed on the window
  │   → assert_true: If this is failing: DANGER, are you sure you want to expose the new interface BigUint64Array to all webpages as a property on the global? Do not make a change to this file without review from jdm or Ms2ger for that specific change! expected true got false
  │ 
  │ test_interfaces/<@http://web-platform.test:8000/_mozilla/mozilla/interfaces.js:73:18
  │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:2024:25
  │ test@http://web-platform.test:8000/resources/testharness.js:548:30
  │ test_interfaces@http://web-platform.test:8000/_mozilla/mozilla/interfaces.js:2:7
  └ @http://web-platform.test:8000/_mozilla/mozilla/interfaces.html:13:16
```

Not going to pop the champagne bottle right now, but that seems promising.

nox commented:
I removed the failing expectations and fixed the interfaces test.

@bors-servo try=wpt

I think this is ready for review.

r? @jdm

bors-servo commented:
:hourglass: Trying commit 333783509cb44af195e045c0e8218b559827627b with merge 1418c0555ff77e5a3d6cf0c6020ba92ece36be2e...

nox commented:
> I think the best way to avoid this problem in the future is to add C wrappers for all of the C++ JSValue inline creation/extraction methods, and write a unit test that verifies that our Rust implementations match for all of the possible cases.

Missed that comment. The point of those Rust-written methods is that the missing inline opportunities were killing us in various benchmarks such as the Dromaeo one.

bors-servo commented:
:broken_heart: Test failed - [status-taskcluster](https://community-tc.services.mozilla.com/tasks/groups/WMNndDP5QMq8SSv1JTqcpQ)

jdm commented:
Yes, I didn't meanwe should use the C wrappers in any code except the test code.

bors-servo commented:
:umbrella: The latest upstream changes (presumably #25369) made this pull request unmergeable. Please resolve the merge conflicts.

nox commented:
Roadmap to land this:

* Land https://github.com/servo/mozjs/pull/225
* Fix Window test failure in https://github.com/servo/rust-mozjs/pull/494
* Land https://github.com/servo/rust-mozjs/pull/494
* Land https://github.com/servo/rust-webvr/pull/97
* Land https://github.com/servo/webxr/pull/125
* Remove the git dependency tweaks in this PR
* Land this PR

nox commented:
Roadmap to land this:

* Land https://github.com/servo/mozjs/pull/225
* Fix Window test failure in https://github.com/servo/rust-mozjs/pull/494
* Land https://github.com/servo/rust-mozjs/pull/494
* Land https://github.com/servo/rust-webvr/pull/97
* Land https://github.com/servo/webxr/pull/125
* Remove the git dependency tweaks in this PR
* Land this PR

nox commented:
@bors-servo try

bors-servo commented:
:hourglass: Trying commit ca6a6200b614440ecaa614b8bea3ef950048fa63 with merge 2d062d4804d5f5101aaef8b79d3eb62db7fca943...

bors-servo commented:
:broken_heart: Test failed - [status-taskcluster](https://community-tc.services.mozilla.com/tasks/groups/GWxqQFMPSbCI17nroSVK8Q)

bors-servo commented:
:umbrella: The latest upstream changes (presumably #25806) made this pull request unmergeable. Please resolve the merge conflicts.

nox commented:
@bors-servo try

bors-servo commented:
:hourglass: Trying commit 4406a9a18184191bf50c7ec84add2b13b5a208a3 with merge 22b2134224ab0bfad4414bbc9285ea112b7ca67e...

bors-servo commented:
:broken_heart: Test failed - [status-taskcluster](https://community-tc.services.mozilla.com/tasks/groups/VWblVo_6T7OPC4FEvXoXYg)

nox commented:
@bors-servo try

bors-servo commented:
:hourglass: Trying commit 10793dd9492f2a097d2eb773a44caffecd257535 with merge 2254db22fd7ed8d9ec1069e43fee034272b1ebec...

bors-servo commented:
:broken_heart: Test failed - [status-taskcluster](https://community-tc.services.mozilla.com/tasks/groups/cWkRVfK6QLma7fodIgt_uA)

jdm commented:
```
INFO: checking for android ndk version... 
ERROR: The only supported version of the NDK is r20 (have r15c)
```
Well, that's just delightful news to receive.

bors-servo commented:
:umbrella: The latest upstream changes (presumably #25810) made this pull request unmergeable. Please resolve the merge conflicts.

nox commented:
@bors-servo try

bors-servo commented:
:hourglass: Trying commit 9fb8cadc19474c79d74d28d3f45b127b480d9b68 with merge 69edc40899b15b47ead20248401a4082f304ce58...

bors-servo commented:
:broken_heart: Test failed - [status-taskcluster](https://community-tc.services.mozilla.com/tasks/groups/aIOAIHvJT3iVEZDP3uzUqg)

nox commented:
Wrote `ENABLE_UWP` instead of `JS_ENABLE_UWP`, ugh.

@bors-servo try

bors-servo commented:
:hourglass: Trying commit 4cc5dc433d8973295799c4854e81e6100dcf3711 with merge 99b4d06dbd9331e63701771dd045bba051295bc8...

bors-servo commented:
:broken_heart: Test failed - [status-taskcluster](https://community-tc.services.mozilla.com/tasks/groups/FrISkMyCTY27-uecpgqhOw)

nox commented:
```
In file included from Unified_cpp_mfbt0.cpp:47:
C:/Users/task_1582559890/.cargo/git/checkouts/mozjs-fa11ffc7d4f1cc2d/f03b72e/mozjs/mfbt/Poison.cpp(49,10): error: use of undeclared identifier 'VirtualAllocFromApp'
  return VirtualAllocFromApp((void*)aRegion, aSize, MEM_RESERVE, PAGE_NOACCESS);
         ^
1 error generated.
```

Either I'm dumb or this function is somehow not found. I patched `mozjs_sys` to properly include `memoryapi.h` when `JS_ENABLE_UWP` is enabled so I don't understand why this is still happening. Will look into that tomorrow.

nox commented:
Is the CI even running on Windows 10?

jdm commented:
It is. Curiously, JS_ENABLE_UWP is not being passed as a define in the command line:
```
c:/Users/task_1582559890/repo/.servo/msvc-dependencies/llvm/8.0.1/bin/clang-cl.exe -FoUnified_cpp_mfbt0.obj -c -guard:cf -DNDEBUG=1 -DTRIMMED=1 -DIMPL_MFBT -DLZ4LIB_VISIBILITY= -IC:/Users/task_1582559890/.cargo/git/checkouts/mozjs-fa11ffc7d4f1cc2d/f03b72e/mozjs/mfbt -IC:/Users/task_1582559890/repo/target/x86_64-uwp-windows-msvc/debug/build/mozjs_sys-d4cdce9349c9bf5c/out/build/mfbt -IC:/Users/task_1582559890/.cargo/git/checkouts/mozjs-fa11ffc7d4f1cc2d/f03b72e/mozjs/mfbt/double-conversion -IC:/Users/task_1582559890/repo/target/x86_64-uwp-windows-msvc/debug/build/mozjs_sys-d4cdce9349c9bf5c/out/build/dist/include -MD -FI C:/Users/task_1582559890/repo/target/x86_64-uwp-windows-msvc/debug/build/mozjs_sys-d4cdce9349c9bf5c/out/build/js/src/js-confdefs.h -DMOZILLA_CLIENT -Qunused-arguments -IC:/Users/task_1582559890/.cargo/registry/src/github.com-1ecc6299db9ec823/encoding_c_mem-0.2.5/include -Qunused-arguments -DWINAPI_FAMILY=WINAPI_FAMILY_APP -MD -DWINAPI_FAMILY=WINAPI_FAMILY_APP -TP -nologo -D_CRT_SECURE_NO_WARNINGS -Zc:sizedDealloc- -D_HAS_EXCEPTIONS=0 -W3 -Gy -Zc:inline -Gw -Wno-inline-new-delete -Wno-invalid-offsetof -Wno-microsoft-enum-value -Wno-microsoft-include -Wno-unknown-pragmas -Wno-ignored-pragmas -Wno-deprecated-declarations -Wno-invalid-noreturn -Wno-inconsistent-missing-override -Wno-implicit-exception-spec-mismatch -Wno-microsoft-exception-spec -Wno-unused-local-typedef -Wno-ignored-attributes -Wno-used-but-marked-unused -GR- -Z7 -O2 -Oy -Xclang -MP -Xclang -dependency-file -Xclang .deps/Unified_cpp_mfbt0.obj.pp -Xclang -MT -Xclang Unified_cpp_mfbt0.obj Unified_cpp_mfbt0.cpp 
```

jdm commented:
Although I guess that's supposed to come from js-config.h instead (https://github.com/servo/mozjs/blob/99952b4e799a9b2d7647176d6f2d94d78e642b7e/mozjs/js/src/js-config.h.in#L58).

jdm commented:
That being said, earlier in the log there is
```
c:/Users/task_1582559890/repo/.servo/msvc-dependencies/llvm/8.0.1/bin/clang-cl.exe -P -Fiself-hosting-preprocessed.pp -Qunused-arguments -IC:/Users/task_1582559890/.cargo/registry/src/github.com-1ecc6299db9ec823/encoding_c_mem-0.2.5/include -DRELEASE_OR_BETA=1 -DHAVE__GETC_NOLOCK=1 -DHAVE_64BIT_BUILD=1 -DCROSS_COMPILE= -D_WIN32_WINNT=0x601 -DHAVE_SYS_TYPES_H=1 -DJS_PUNBOX64=1 -DMOZILLA_UAVERSION="72.0" -DNO_NSPR_10_SUPPORT=1 -DJS_64BIT=1 -DHAVE_STDINT_H=1 -DHAVE_FORCEINLINE=1 -DMOZILLA_VERSION="72.0" -DHAVE_INTTYPES_H=1 -D_USE_MATH_DEFINES=1 -DNOMINMAX=1 -DMALLOC_H=<malloc.h> -DSTDC_HEADERS=1 -DNO_RUST_PANIC_HOOK=1 -DWIN32_LEAN_AND_MEAN=1 -DNDEBUG=1 -DHAVE_ALIGNED_MALLOC=1 -DJS_STANDALONE=1 -DMOZ_BUILD_APP=js -DX_DISPLAY_MISSING=1 -D_CRT_NONSTDC_NO_WARNINGS=1 -DMOZJS_MINOR_VERSION=0 -DWIN32=1 -DTRIMMED=1 -DHAVE_LOCALECONV=1 -D_WIN32_IE=0x0800 -DMOZ_STATIC_JS=1 -DMOZ_DLL_PREFIX="" -DHAVE_MALLOC_H=1 -DWINVER=0x601 -DMOZILLA_VERSION_U=72.0 -D_CRT_SECURE_NO_WARNINGS=1 -DSTATIC_EXPORTABLE_JS_API=1 -DHAVE_CPUID_H=1 -DJS_DEFAULT_JITREPORT_GRANULARITY=3 -DJS_ENABLE_UWP=1 -DMOZ_DLL_SUFFIX=".dll" -DMOZ_UPDATE_CHANNEL=default -DJS_WITHOUT_NSPR=1 -DXP_WIN=1 -DMALLOC_USABLE_SIZE_CONST_PTR=const -D_AMD64_=1 -DMOZJS_MAJOR_VERSION=72 -DJS_CODEGEN_X64=1 -DHAVE__MSIZE=1 -D_WINDOWS=1 self-hosting-cpp-input.cpp
```
which includes a `-DJS_ENABLE_UWP=1`.

bors-servo commented:
:umbrella: The latest upstream changes (presumably #25796) made this pull request unmergeable. Please resolve the merge conflicts.

nox commented:
@jdm I doubt `JS_ENABLE_UWP` isn't set: the function that the compiler can't found is called in an `#ifdef JS_ENABLE_UWP` block.

nox commented:
> That being said, earlier in the log there is

That command with `-DJS_ENABLE_UWP=1` is for some generated stuff that isn't the common path for building C++ code in SM.

nox commented:
AFAICT, we do not include `js-config.h` when building MFBT, we use `js-confdefs.h`, which seems to only include configure variables from the `old-configure` script, not the fancy Python ones we define ourselves.

nox commented:
`add_old_configure_assignment` seems relevant.

nox commented:
Rebased, tweaked the setting of `JS_ENABLE_UWP` and fixed a syntax error somewhere in `DynamicallyLinkedFunctionPtr.h`, let's see how it goes this time.

@bors-servo try

bors-servo commented:
:hourglass: Trying commit 5a16305015f8a9b82fbc876409b71d2ab05b9efa with merge 53a851c59c98fcb4c4967c9be9320d02c237cfab...

bors-servo commented:
:broken_heart: Test failed - [status-taskcluster](https://community-tc.services.mozilla.com/tasks/groups/CRoZHyaWSAadVTkfoKw8YA)

nox commented:
Seems like the syntax error fix brought some progress to "Servo: Windows x64: Dev build + unit tests", but I still have that `JS_ENABLE_UWP` issue.

nox commented:
Wondering if we just don't target Windows 10 when building MFBT specifically.

nox commented:
@bors-servo try

Fixed another syntax error (GCC-specific variadic macro syntax).

bors-servo commented:
:hourglass: Trying commit 911bf06589a4d3d313972a965a193edd7bf8ae78 with merge 80ba8edb60156f9fc7566f2729c77eb7deac1d87...

bors-servo commented:
:broken_heart: Test failed - [status-taskcluster](https://community-tc.services.mozilla.com/tasks/groups/Zm8l5AcWQgSjdNAx1CaD6A)

nox commented:
Aha, we patch `old-configure.in` to target Windows 10, but we don't patch `old-configure`.

@bors-servo try

bors-servo commented:
:hourglass: Trying commit 6d801a3d5d0ff13ee18afa57a58e9df29fee06e1 with merge 6479836d7a7a908b129ddd7107e7f2d9a1f33110...

bors-servo commented:
:broken_heart: Test failed - [status-taskcluster](https://community-tc.services.mozilla.com/tasks/groups/SKZvS6FCRim9ZQ912-d14w)

nox commented:
Oh yeah, there is still this issue:

> ERROR: Don't know how to translate aarch64-windows-mingw32 for rustc

Will look into it after lunch and 1:1.

nox commented:
Build "Windows x64: Dev build + unit tests" is failing in the compilation of the glue, which means that SM at least built successfully.

Build "Windows x64: UWP dev build" is failing with the following errors, while building SM itself:

```
C:/Users/task_1582621701/.cargo/git/checkouts/mozjs-fa11ffc7d4f1cc2d/f3d5005/mozjs/mozglue/misc/WindowsUnicode.cpp(50,39): error: unknown type name 'PCUNICODE_STRING'; did you mean '_UNICODE_STRING'?
C:/Users/task_1582621701/.cargo/git/checkouts/mozjs-fa11ffc7d4f1cc2d/f3d5005/mozjs/mozglue/misc/WindowsUnicode.cpp(50,56): error: variable has incomplete type '_UNICODE_STRING'
```

nox commented:
WindowsUnicode.cpp includes winternl.h, which defines a single desktop-only UNICODE_STRING version. I suspect that for UWP, it doesn't even bothering including the header that actually defines UNICODE_STRING, which is actually ntdef.h.

nox commented:
AGAIN

@bors-servo try

bors-servo commented:
:hourglass: Trying commit 11cb89c63070eac86df87fcd26d256676c69e7ae with merge 582a52dd769300e2b4085e3f0f7c6166dc0eb9ad...

bors-servo commented:
:broken_heart: Test failed - [status-taskcluster](https://community-tc.services.mozilla.com/tasks/groups/Lk8w6kYcSNaxUAu7ls8y8g)

nox commented:
@retep998 What is the header where `UNICODE_STRING` and `PCUNICODE_STRING` are defined?

nox commented:
I just gated the symbol with `ifndef JS_ENABLE_UWP` because this isn't even used anyway from the tarball.

nox commented:
I just gated the symbol with `ifndef JS_ENABLE_UWP` because this isn't even used anyway from the tarball.

@bors-servo try

bors-servo commented:
:hourglass: Trying commit 998290e93e1218c360988d3861df10a957cec949 with merge 510ee7d876bb62430246f650491a60069d16477f...

bors-servo commented:
:broken_heart: Test failed - [status-taskcluster](https://community-tc.services.mozilla.com/tasks/groups/KJlvW70ZQXy6_jjes3Evmg)

retep998 commented:
@nox https://docs.rs/winapi/0.3.8/winapi/shared/ntdef/struct.UNICODE_STRING.html

nox commented:
Still 2 broken Windows builds:

* the triple error with aarch64-windows-mingw32;
* jsglue.cpp is compiled with cl instead of clang-cl and with different flags than SM.

nox commented:
Passing `-vv` to `mach build` on the CI to make it print the env just before running `cargo build`.

nox commented:
@bors-servo try

bors-servo commented:
:hourglass: Trying commit d5347934968b104e97971b84ee990db8c09a09a1 with merge 4c25e02f26f7536edbf23a360d56604fb9507378...

bors-servo commented:
:broken_heart: Test failed - [status-taskcluster](https://community-tc.services.mozilla.com/tasks/groups/SZjQ7qLQTrOGJgcjbdLxcQ)

nox commented:
@bors-servo try

bors-servo commented:
:hourglass: Trying commit b739fdcf3ef2a7886c2cb99f8c74d74a81ae21ea with merge 941458b22346a458e06a6a6050fc5ad8e1e385a5...

bors-servo commented:
:broken_heart: Test failed - [status-taskcluster](https://community-tc.services.mozilla.com/tasks/groups/XdGkJ6vGQvG-cG53Baj8BQ)

bors-servo commented:
:umbrella: The latest upstream changes (presumably #25843) made this pull request unmergeable. Please resolve the merge conflicts.

nox commented:
The triplet issue is https://bugzilla.mozilla.org/show_bug.cgi?id=1595218.

nox commented:
@bors-servo try

bors-servo commented:
:hourglass: Trying commit 407a7a0b41aa15511e43b457a07c7e5c92d731c1 with merge 947a70bd1b44d980563d7344036f3680b9fa55de...

bors-servo commented:
:broken_heart: Test failed - [status-taskcluster](https://community-tc.services.mozilla.com/tasks/groups/SijJR9LJR6eCM0DWCiIECQ)

nox commented:
@bors-servo try

bors-servo commented:
:hourglass: Trying commit a4440dc727711255d1f049c1b85fcaba34eac7f3 with merge 4b324daefe89eeab2dfda3952638f8ba25fb0907...

bors-servo commented:
:broken_heart: Test failed - [status-taskcluster](https://community-tc.services.mozilla.com/tasks/groups/CtwE_Z0wS1KJutdgPaEztQ)

bors-servo commented:
:hourglass: Trying commit 379bdd36d79f0c3c38c26fb2407e9b7699d121fd with merge 730533590b63b55cf560964a999345681b13b0ff...

bors-servo commented:
:broken_heart: Test failed - [status-taskcluster](https://community-tc.services.mozilla.com/tasks/groups/cKnG-UDyRL2u1mZb_WaJeQ)

nox committed:
Update SpiderMonkey

nox commented:
@emilio I updated bindgen to 0.53, which means we may need `.size_t_is_usize(true)` on the Gecko side of Stylo.

bors-servo commented:
:umbrella: The latest upstream changes (presumably #25863) made this pull request unmergeable. Please resolve the merge conflicts.

asajeffrey commented on Cargo.toml:
```diff
@@ -32,3 +32,10 @@ mio = { git = "https://github.com/servo/mio.git", branch = "servo" }
+rust-webvr-api = { git = "https://github.com/servo/rust-webvr.git", branch = "bindgen" }
+[patch."https://github.com/servo/webxr"]
+webxr = { git = "https://github.com/servo/webxr.git?lol", branch = "bindgen" }
+webxr-api = { git = "https://github.com/servo/webxr.git?lol", branch = "bindgen" }
```
Why do we need to update webvr/xr?

nox replied:
For bindgen.

asajeffrey commented on components/script/dom/bindings/codegen/CodegenRust.py:
```diff
@@ -5950,10 +5968,10 @@ def generate_imports(config, cgthings, descriptors, callbacks=None, dictionaries
+        'js::jsapi::JSPropertySpec_Accessor',
+        'js::jsapi::JSPropertySpec_AccessorsOrValue',
+        'js::jsapi::JSPropertySpec_AccessorsOrValue_Accessors',
+        'js::jsapi::JSPropertySpec_Name',
```
More yay for bindgenup!

asajeffrey reviewed (commented):
LGTM. You can r=me once the other changes land.

nox commented:
@bors-servo r=asajeffrey p=1

bors-servo commented:
:pushpin: Commit 5a4f8cf has been approved by `asajeffrey`

<!-- @bors-servo r=asajeffrey 5a4f8cf93f9f674a164a0a3cfc586accef3d06f9 -->

bors-servo commented:
:hourglass: Testing commit 5a4f8cf93f9f674a164a0a3cfc586accef3d06f9 with merge 356c4e0bc860307ca373f85156e816f8c0a0d132...

CYBAI commented on components/script/dom/globalscope.rs:
```diff
@@ -2158,15 +2158,20 @@ impl GlobalScope {
-                        code.as_ptr() as *const _,
-                        code.len() as libc::size_t,
+                        &mut SourceText {
+                            units_: code.as_ptr() as *const _,
```
This makes me notice I forgot to file an issue to expose this function from rust-mozjs 👀 Let me file one!

https://github.com/servo/servo/blob/ca25e18791163de5c6af1267d876c7f5bb705f61/components/script/script_module.rs#L78-L85

CYBAI replied:
Ah, nah, I did file one at servo/rust-mozjs#490. Will send a PR to fix it and update here and `script_module`.

bors-servo commented:
:sunny: Test successful - [status-taskcluster](https://community-tc.services.mozilla.com/tasks/groups/TOEnuu5mRdWdtpRLilGBoQ)
Approved by: asajeffrey
Pushing 356c4e0bc860307ca373f85156e816f8c0a0d132 to master...
